### PR TITLE
perf: added possibility to lazy initialize the document

### PR DESCRIPTION
**Pull request by H4ad**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?

Currently, NestJS builds the Swagger document synchronously on startup, which it's a waste of time since we don't care about the swagger routes until we need to make a call to it.

## What is the new behavior?

So, I added an option to lazy initialize the Swagger document, the document will only be built when we have some request for it.

In my API, it reduced about 70~75ms from initialization, it will vary according to the API and the amount of routes but now it will only take 2ms to initialize the Swagger Module.

### Usage

Now, to use the new lazy loading feature, just change these lines:

```diff
const swaggerDocument = new DocumentBuilder()
  .setTitle(config.get<string>('SWAGGER_TITLE') ?? 'Base')
  .setDescription(config.get<string>('SWAGGER_DESCRIPTION') ?? 'The base API')
  .setVersion(config.get<string>('VERSION') ?? '1.0.0')
  .addTag('main')
  .addBearerAuth()
  .build();

-const document = SwaggerModule.createDocument(app, swaggerDocument);
+const document = () => SwaggerModule.createDocument(app, swaggerDocument);

SwaggerModule.setup('/swagger', app, document, {
  swaggerOptions: { docExpansion: 'none' },
});
```

Just one line to use it.

### Reproduction

To use it: https://stackblitz.com/edit/nestjs-starter-bql4af?file=src/main.ts

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by micalevisk** [comment]
A good improvement even for that small app :tada: 

<table>
  <tr>
    <th>eager</th>
    <th>lazy</th>
  </tr>
  <tr>
    <td valign="top">before
</td>
    <td valign="top">after</td>
  </tr>
</table>


--- 

I wonder if the API couldn't be this instead:

```ts
const lazyDocument = () => SwaggerModule.createDocument(app, swaggerDocument)
SwaggerModule.setup('/swagger', app, lazyDocument, {
  swaggerOptions: { docExpansion: 'none' }
})
```

**Comment by H4ad** [comment]
@micalevisk Yeah, I think your way could be better since I don't need to add more options to `SwaggerCustomOptions`.

**Comment by H4ad** [comment]
Updated with the new version, it was cleaner than the older version.

**Comment by H4ad** [comment]
Docs PR: https://github.com/nestjs/docs.nestjs.com/pull/2704

File: lib/swagger-module.ts [hunk 1]
```diff
@@ -123,33 +161,46 @@ export class SwaggerModule {
...
     httpAdapter.get(normalizeRelPath(options.yamlDocumentUrl), (req, res) => {
       res.type('text/yaml');
-      res.send(options.yamlDocument);
+
+      if (!document)
+        document = lazyBuildDocument();
+
+      if (!yamlDocument)
+        yamlDocument = jsyaml.dump(document);
+
```
Comment (Tony133) [inline lib/swagger-module.ts:186]:
```suggestion
      if (!document) {
        document = lazyBuildDocument();
      }
      
      if (!yamlDocument) {
         yamlDocument = jsyaml.dump(document);
      }

```

File: lib/swagger-module.ts [hunk 2]
```diff
@@ -123,33 +161,46 @@ export class SwaggerModule {
...
     httpAdapter.get(normalizeRelPath(options.jsonDocumentUrl), (req, res) => {
       res.type('application/json');
-      res.send(options.jsonDocument);
+
+      if (!document)
+        document = lazyBuildDocument();
+
+      if (!yamlDocument) jsonDocument = JSON.stringify(document);
+
```
Comment (Tony133) [inline lib/swagger-module.ts:174]:
```suggestion
      if (!document) {
         document = lazyBuildDocument();
      }
      
      if (!yamlDocument) { 
         jsonDocument = JSON.stringify(document);
      }
```

File: lib/swagger-module.ts [hunk 3]
```diff
@@ -105,14 +129,28 @@ export class SwaggerModule {
...
     try {
       httpAdapter.get(normalizeRelPath(`${finalPath}/`), (req, res) => {
         res.type('text/html');
-        res.send(options.html);
+
+        if (!document)
+          document = lazyBuildDocument();
+
+        if (!html)
+          html = buildSwaggerHTML(baseUrlForSwaggerUI, document, options);
```
Comment (Tony133) [inline lib/swagger-module.ts:151]:
```suggestion
        if (!document) {
          document = lazyBuildDocument();
        }
        
        if (!html) {
          html = buildSwaggerHTML(baseUrlForSwaggerUI, document, options);
        }
```

File: lib/swagger-module.ts [hunk 4]
```diff
@@ -105,14 +129,28 @@ export class SwaggerModule {
 
     httpAdapter.get(finalPath, (req, res) => {
       res.type('text/html');
-      res.send(options.html);
+
+      if (!document)
+        document = lazyBuildDocument();
+
+      if (!html)
+        html = buildSwaggerHTML(baseUrlForSwaggerUI, document, options);
```
Comment (Tony133) [inline lib/swagger-module.ts:137]:
```suggestion
      if (!document) {
        document = lazyBuildDocument();
      }
      
      if (!html) {
        html = buildSwaggerHTML(baseUrlForSwaggerUI, document, options);
      }
```

File: lib/swagger-module.ts [hunk 5]
```diff
@@ -89,12 +106,19 @@ export class SwaggerModule {
...
+          `${finalPath}/${urlLastSubdirectory}/swagger-ui-init.js`,
         ),
         (req, res) => {
           res.type('application/javascript');
+
+          if (!document)
+            document = lazyBuildDocument();
+
+          if (!swaggerInitJS)
+            swaggerInitJS = buildSwaggerInitJS(document, options);
```
Comment (Tony133) [inline lib/swagger-module.ts:118]:
```suggestion
          if (!document) {
            document = lazyBuildDocument();
          }

          if (!swaggerInitJS) {
            swaggerInitJS = buildSwaggerInitJS(document, options);
          }
```

File: lib/swagger-module.ts [hunk 6]
```diff
@@ -65,21 +66,37 @@ export class SwaggerModule {
...
       normalizeRelPath(`${finalPath}/swagger-ui-init.js`),
       (req, res) => {
         res.type('application/javascript');
+
+        if (!document)
+          document = lazyBuildDocument();
+
+        if (!swaggerInitJS)
+          swaggerInitJS = buildSwaggerInitJS(document, options);
+
```
Comment (Tony133) [inline lib/swagger-module.ts:97]:
```suggestion
        if (!document) {
          document = lazyBuildDocument();
        }
        
        if (!swaggerInitJS) {
          swaggerInitJS = buildSwaggerInitJS(document, options);
        }

```

**Review by Tony133 — COMMENTED** [review]
Just some minor corrections for code identation, for the rest is ok 👍

**Review by micalevisk — APPROVED** [review]

**Review by rubiin — APPROVED** [review]

**Comment by H4ad** [comment]
Fixed issues with e2e tests!

**Review by rubiin — APPROVED** [review]

**Review by Tony133 — APPROVED** [review]

**Comment by kamilmysliwiec** [comment]
It seems that the PR caused this regression https://github.com/nestjs/swagger/issues/2481

**Pull request merged**